front-end/src/App.js

```diff
@@ -1,29 +1,56 @@
-import React from "react";
-import Profile from "./Pages/Profile.js";
+import React, { useState, useMemo } from "react";
+import Account from "./Pages/Account.js";
 import Home from "./Pages/Home.js";
 import Navbar from "./Components/Navbar.js";
 import Sidebar from "./Components/Sidebar.js";
 import { HashRouter, Route } from "react-router-dom";
+import { UserContext } from "./UserContext.js";
+import Login from "./Pages/Login";
+import Profile from "./Pages/Profile";
 
 export default function App() {
+  const [user, setUser] = useState(null);
+
+  const providerValue = useMemo(() => ({ user, setUser }), [user, setUser]);
+
   return (
     <HashRouter basename="/">
```

Comment:
I'd use BrowserRouter instead of HashRouter

Reply:
one hash router won't keep your browser history correct so you'll get things like

profile/1/#/about/profile

Reply:
it gives the illusion of url protection but it really is just forcing you to do extra logic for it.

Reply:
you don't want a user to be on `/account/1` and be shown a login page.

you want them redirected to the login page.

Reply:
Okay, I am using browser router now. if the user is logged in they will be shown their profile if they are on the route /myprofile/number. but if they are not logged in and they somehow get to that route, the login page will be rendered instead.